### PR TITLE
Fixed redirect bug in partner sites

### PR DESCRIPTION
**Pull request by smclay**
There was a bug in the partner sites feature. It was largely down to a misunderstanding on my part. I thought the partner site server should act as a proxy but instead we should stream the file directly from where it is hosted.
I also added a new function to parse the host ID. Perhaps one already exists. If so I can remove it and reuse the existing function.

### CODE REVIEW

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py [hunk 1]
```diff
@@ -73,6 +73,33 @@
...
+    """
+    try:
+        host = host_id.split(":")[0]
+        domain = None
+        if "." in host:
+            domain = host.split(".", 1)[-1]
+        port = int(host_id.split(":")[-1])
+        return host, domain, port
+    except IndexError:
+        return None, None, None
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/ngamsFileUtils.py:100]: `string.split` always produces a list with at least one element, so indexing with `0` and `-1` will always work, and no `IndexError` should occur.

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py [hunk 2]
```diff
@@ -154,8 +180,13 @@ def lookup_partner_site_file_status(ngas_server,
...
             disk_info = status_info.getDiskStatusList()[0]
             file_info = disk_info.getFileObjList()[0]
-            host = partner_address
-            port = partner_port
+            # This is a bit of a hack because the host ID may not contain
+            # the fully qualified address. We append the domain name from
+            # the partner site address. This should work because they are
+            # part of the same cluster.
+            host, domain, port = parse_host_id(disk_info.getHostId())
+            if domain is None:
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/ngamsFileUtils.py:188]: You need to check here if `partner_domain` for `None`, this is what's causing the tests to fail.

**Review by rtobar — COMMENTED** [review]
Hi @smclay, sorry for the initial faillures on the tests again, this PR prompted me to have a look at them and now the `master` is back to green, as you probably saw already. After that there is still a unit test failure which looks genuine. We sadly don't have a very stable coverage (it still oscillates within +-0.5% or so) so in principle don't pay much attention to the "coverage decreased" messages.

> I thought the partner site server should act as a proxy but instead we should stream the file directly from where it is hosted.

I'm a bit puzzled by this statement, because the changes in this PR don't change that behavior; that is, the file is still proxied from the partner site through the server originally contacted by the client. When I read this statement I thought you'd try to send a redirect response to the client for them to contact the partner site directly. Is that your intent? Otherwise the changes in this PR only affect how a server contacts its partner site when doing a request.

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py (hunk 1 above)
Comment (smclay) [reply src/ngamsServer/ngamsServer/ngamsFileUtils.py]: OK. I have removed the exception handling

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py (hunk 2 above)
Comment (smclay) [reply src/ngamsServer/ngamsServer/ngamsFileUtils.py]: Yes. You are correct. I have resolved this.

**Comment by smclay** [comment]
@rtobar to answer your question. The way partner sites feature should work is...
1. NGAS receives a retrieve request
2. NGAS looks in the local host and cluster but fails to find the request file
3. NGAS sends a status request for the requested file to the nominated partner site (from config)
4. Partner site response confirms that file is available in the partner site cluster (with host ID where file is hosted)
5. NGAS sends a retrieve request to the partner site server where the file is hosted
6. NGAS streams the file data from the partner site server to the original client that sent the request

The mistake I made was to send the retrieve request to the nominated partner site instead of the partner site host where the file resides. This results in a redirect response. This is not what we want.

I have applied a bug fix to my latest bug fix. I still get problems running the test_partner_site.py unit test on my VM. The problem is that ngamsHttpUtils struggles with the host name. For example, it can handle requests to localhost but not to the host name of my VM. I have checked that my VM is properly configured. The CURL client works fine. Hopefully this passes on your test server.

**Comment by rtobar** [comment]
@smclay Thanks for the clarification, I had misunderstood what was meant by "directly from where it is hosted". I thought this meant that it had to be a direct transfer from the standpoint of the client, but from your explanation it became clear it means direct from the standpoint of the server the client contacts first; the file is still streamed through this first server, which is why I was confused.

The failures in the test seem genuine. The problem has to do with the difference between IPs and names -- by default the servers spawned by unit tests listen only in `127.0.0.1`, but your code is now trying to contact the partner site via its hostname, which resolves to the IP of one of the other interfaces, and fails. You can overcome this by starting the partner site server with `NgamsCfg.Server[1].IPAddress` = `0.0.0.0` (see https://ngas.readthedocs.io/en/master/configuration.html#server), which should allow you to contact the partner site through any of its interfaces.

**Comment by smclay** [comment]
@rtobar Thank you for the suggestion. It fixed the problem. I have bumped up against this Server->IpAddress before. Would it not make more sense to set the default to 0.0.0.0? Hopefully now the bug fixes can be merged. Let me know if there is anything else I need to do.

**Comment by smclay** [comment]
@rtobar I am struggling to fix this test. It is working on my local machine. I think the problem might be the configuration of the travis host machine.

I noticed in the build log that the default host name is `b2b6d956-bab4-4178-aba9-2af50295ff6e@1.worker-com-oss-794dcff96c-xtpdn.gce-production-3`. I suspect this host name would cause problems in the test because `ngas_hosts.host_id` columns size is too short to store the full host name.

There is a command in the build script to change the host name. The command is `sudo hostname my-ngas-host`. This looks incorrect to me but then again it appears the build is running on a very old version of xenial linux which is derived from `Ubuntu 16.04.6 LTS`. I would normally expect the command to be `sudo hostnamectl set-default my-ngas-host`. But perhaps systemd is not used in xenial linux. What the domain name is anybody's guess.

Is it possible to change the host OS used for the build test? Could we change to CentOS7?

**Comment by smclay** [comment]
I decided to set `localhost` everywhere. This works but it is a bit of a work around solution. It would be preferable to use real host name values. Anyway, I need this branched to get merged so I can build new RPM packages and carry on with our end to end testing.

**Pull request merged**